Import Jupyter Notebook files into Atom

kylebarron opened the pull request:
This PR adds functionality to import Jupyter Notebook files as text files in Atom.

It imports the new file into a new TextEditor. In order to support Markdown blocks, I need to know the correct comment symbol, so I need to find the correct Atom Grammar for the notebook file. I first try to match file extensions between the notebook file's metadata and the file extensions that each Grammar is applied for. But since the file extension is an optional metadata field for the notebook, I also attempt to match on kernelspec name. These two together should work for the vast majority of notebooks.

Given the grammar, I get the source text for code and markdown cells, prepending each line of text in the markdown cell with the `commentStartString`, right trimmed plus a space. I add a cell marker line before each cell, either `# %%` or `# %% markdown`. It uses `\r\n` as the line separator on Windows; `\n` otherwise (for newlines created between cells).

To do:

- I want to add tests, especially testing that the round-trip import-export and export-import from Hydrogen is idempotent, but that depends on the result of #1498 and #1499. Regardless, if there are any substantial requested changes to this PR, I want to make the changes before making the tests.

Limitations:

- Currently only version 4 of the notebook format is supported.
- Only imports source text of code cells and Markdown cells. Doesn't import raw cells or code cell outputs.

### Python
![image](https://user-images.githubusercontent.com/15164633/50504254-8a324d00-0a29-11e9-9937-403836328664.png)
```py
# %% markdown
# # Example Notebook
# 
# This is an example Python Notebook!
# %%
print('hello world!')
# %%
```

### Bash:
![image](https://user-images.githubusercontent.com/15164633/50504202-30318780-0a29-11e9-97b7-d76867d0eac3.png)
```bash
# %% markdown
# # Example Notebook
# 
# This is an example Bash Notebook.
# %%
echo "hello world!"
# %%
```

### Javascript:
![image](https://user-images.githubusercontent.com/15164633/50504222-4c352900-0a29-11e9-8658-7dee6aaba6f2.png)

```js
// %% markdown
// # Example Notebook
// 
// This is an example Javascript Notebook.
// %%
console.log('hello world!');
// %%
```

### R:
![image](https://user-images.githubusercontent.com/15164633/50504233-67a03400-0a29-11e9-9148-fc6ee6202909.png)
```r
# %% markdown
# # Example Notebook
# 
# This is an example R Notebook!
# %%
print('hello world!')
# %%
```


Closes #1457, ref #1404.

kylebarron commented:
Can someone help with this last flow error? My understanding of promises and callbacks is only so-so.

```
Cannot call readFile with loadNotebook bound to callback because:
 • Promise [1] is incompatible with undefined [2] in the return value.
 • Promise [3] is incompatible with undefined [2] in the return value.

     lib/import-notebook.js
      23│       atom.notifications.addError("Selected file must have extension .ipynb");
      24│       return;
      25│     }
      26│     readFile(filename, loadNotebook);
      27│   });
      28│ }
      29│
 [3]  30│ async function loadNotebook(err, data) {

     /private/tmp/flow/flowlib_f7461a8/core.js
 [1] 612│ declare class Promise<+R> {

     /private/tmp/flow/flowlib_f7461a8/node.js
 [2] 967│     callback: (err: ?ErrnoError, data: Buffer) => void
```

lgeiger commented:
Thanks for getting this started 🎉 

I guess the flow error is because the callback in `fs.readFile` shouldn't be a `async` function. I guess wrapping it into an arrow function might work around this issue. 

I can review the PR in detail next week.

kylebarron commented:
> I added some tests and fixed a couple things on my fork. If you add me as a collaborator on your fork I can push my changes to this PR. Or, I can pr against your fork if you prefer.

Added as contributor

BenRussert commented:
We can rebase last once we are ready to merge. Try this branch out and see if you can find anything that still needs work. I'll take another look during the week as well to see what's left.

kylebarron commented:
I think all your edits make sense, though I haven't tested them in Atom yet.

BenRussert commented on spec/import-notebook-spec.js:
```diff
@@ -0,0 +1,39 @@
+      fail(e);
+      done();
+    });
+  };
```
Definitely meant to 😂

kylebarron commented:
I'd be in favor of moving `importNotebook` to an `atom-workspace` command, so that a text editor doesn't have to be currently active to use it.

kylebarron commented:
I'd also be happy to add documentation to https://nteract.gitbooks.io/hydrogen/docs/Usage/NotebookFiles.html. (When #1498 was merged, the documentation was updated despite that code not being released yet).

kylebarron commented:
I added some more documentation and moved `import-notebook` to be a workspace command.

kylebarron commented:
@BenRussert 	
I think this is probably good to merge if you're happy with it.

I wasn't sure if this should be its own PR, since it might need more discussion, but it would also be useful to add the import notebook functionality as an opener for `.ipynb` files. It's a [simple 10-line change.](https://github.com/kylebarron/hydrogen/commit/34cc5dfe42a1915e486638c2cb09d64be796ba62)

kylebarron commented:
> I know nothing about how this works, but the ideal for me would be to have atom open .ipynb files like this, then export them on save. Everyone around me works directly in jupyter, this way I could work with them pretty seamlessly.

We currently don't support automatic exporting because of the potential for unintentionally overwriting data.

> On a side note, how about having this import/export to/from the rich document format? Even if it only supported importing one language and markdown cells it would be nice to have the choice of which format to work in.

Do you mean markdown documents?

JohnCHarrington commented:
Fair enough, I can add a hotkey for export anyway. 

Yes I mean markdown documents, then the grammar/kernelspec would be used to set the language on the code blocks.

kylebarron commented:
> Fair enough, I can add a hotkey for export anyway.

Well currently when you export a file to a notebook, it brings up the system file selector, so not sure how much time a hotkey would save. Hydrogen has no way to know ahead of time how to name the outputted notebook file.

> Yes I mean markdown documents, then the grammar/kernelspec would be used to set the language on the code blocks.

Markdown documents are more prone to losing some metadata, in particular cell boundaries. It's probably not too difficult to write, but not my focus at the moment.

kylebarron commented:
@JohnCHarrington 
You may also want to check out the newest release of Pandoc https://github.com/jgm/pandoc/releases/tag/2.6
